client/requester.go

```diff
@@ -0,0 +1,292 @@
+		if err == nil || method != "GET" {
+			break
+		}
+		select {
+		case <-retry.C:
+			continue
+		case <-timeout:
+		}
+		break
+	}
+	if err != nil {
+		return nil, err
+	}
+	return rsp, nil
+}
+
+func (br *BaseRequester) Do(ctx context.Context, opts *RequestOptions, result interface{}) (*RequestResponse, error) {
+	// Is the result expecting a websocket?
+	if ws, ok := result.(*Websocket); ok {
+		dialer := websocket.Dialer{
+			NetDial:          br.transport.Dial,
+			Proxy:            br.transport.Proxy,
+			TLSClientConfig:  br.transport.TLSClientConfig,
+			HandshakeTimeout: 5 * time.Second,
+		}
+		conn, _, err := dialer.DialContext(ctx, opts.Path, nil)
+		if err != nil {
+			return nil, err
+		}
+		*ws = conn
+		return nil, nil
+	}
+
+	// Is the result expecting a caller managed body reader?
+	if bodyReader, ok := result.(*BodyReader); ok {
+		httpResp, err := br.raw(ctx, opts.Method, opts.Path, opts.Query, opts.Headers, opts.Body)
+		if err != nil {
+			return nil, err
+		}
+
+		*bodyReader = httpResp.Body
+		return nil, nil
+	}
+
+	// This is a normal sync or async server request
+	httpResp, err := br.rawWithRetry(ctx, opts.Method, opts.Path, opts.Query, opts.Headers, opts.Body)
+	if err != nil {
+		return nil, err
+	}
+	defer httpResp.Body.Close()
+
+	// Get the client decoder to extract what it needs before we proceed
+	reqResp, err := br.decoder(ctx, httpResp, opts, result)
+	if err != nil {
+		return nil, err
+	}
+
+	// Sanity check sync and async requests
+	if opts.Async == false {
+		if reqResp.Type != "sync" {
+			return nil, fmt.Errorf("expected sync response, got %q", reqResp.Type)
+		}
+	} else {
+		if reqResp.Type != "async" {
+			return nil, fmt.Errorf("expected async response for %q on %q, got %q", opts.Method, opts.Path, reqResp.Type)
+		}
+		if reqResp.StatusCode != 202 {
+			return nil, fmt.Errorf("operation not accepted")
+		}
+		if reqResp.Change == "" {
+			return nil, fmt.Errorf("async response without change reference")
+		}
+	}
+
+	return reqResp, nil
+}
+
+// SocketNotFoundError is the error type returned when the client fails
+// to find a unix socket at the specified path.
+type SocketNotFoundError struct {
+	// Err is the wrapped error.
+	Err error
+
+	// Path is the path of the non-existent socket.
+	Path string
+}
+
+func (s SocketNotFoundError) Error() string {
+	if s.Path == "" && s.Err != nil {
+		return s.Err.Error()
+	}
+	return fmt.Sprintf("socket %q not found", s.Path)
+}
+
+func (s SocketNotFoundError) Unwrap() error {
+	return s.Err
+}
+
+// decodeWithNumber decodes input data using json.Decoder, ensuring numbers are preserved
+// via json.Number data type. It errors out on invalid json or any excess input.
+func decodeWithNumber(r io.Reader, value interface{}) error {
+	dec := json.NewDecoder(r)
+	dec.UseNumber()
+	if err := dec.Decode(&value); err != nil {
+		return err
+	}
+	if dec.More() {
+		return fmt.Errorf("cannot parse json value")
```

Comment:
Minor nitpick, since we also capitalize "URL" in error messages (:
```suggestion
		return fmt.Errorf("cannot parse JSON value")
```

Reply:
Nitpick aside, if I understand correctly this error is returned in the case we successfully decoded a JSON response but there's extra text after a valid JSON string. Shouldn't we return a more specific error?
```suggestion
		return fmt.Errorf("cannot decode extra data after JSON response")
```

Reply:
Existing code. Will change if needed.